crates/neon/src/instance/mod.rs

```diff
@@ -0,0 +1,67 @@
+use std::any::Any;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use once_cell::sync::OnceCell;
+
+use crate::context::Context;
+use crate::lifecycle::InstanceData;
+
+static COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+fn next_id() -> usize {
+    COUNTER.fetch_add(1, Ordering::SeqCst)
+}
+
+/// A cell that can be used to allocate data that is global to an instance
+/// of a Neon addon.
+#[derive(Default)]
+pub struct Global<T> {
+    _type: PhantomData<T>,
+    id: OnceCell<usize>,
+}
+
+impl<T> Global<T> {
+    pub const fn new() -> Self {
+        Self {
+            _type: PhantomData,
+            id: OnceCell::new(),
+        }
+    }
+
+    fn id(&self) -> usize {
+        *self.id.get_or_init(next_id)
+    }
+}
+
+impl<T: Any + Send + 'static> Global<T> {
+    pub fn get<'cx, 'a, C>(&self, cx: &'a mut C) -> Option<&'cx T>
+    where
+        C: Context<'cx>,
+    {
+        let r: Option<&T> = InstanceData::globals(cx)
+            .get(self.id())
+            .as_ref()
+            .map(|boxed| boxed.downcast_ref().unwrap());
+
+        unsafe {
+            std::mem::transmute::<Option<&'a T>, Option<&'cx T>>(r)
+        }
+    }
+
+    pub fn get_or_init<'cx, 'a, C, F>(&self, cx: &'a mut C, f: F) -> &'cx T
```

Comment:
`get_or_try_init` is also useful.

If we wanted to be lazy and directly expose the full expressive power of `OnceCell` we could make `Global::<T>::get` return [`&'cx once_cell::unsync::OnceCell<T>`](https://docs.rs/once_cell/latest/once_cell/unsync/struct.OnceCell.html) instead of `Option<&'cx T>`.

Instead of `Vec<Option<Box<Any>>>` we would store `Vec<Box<Any>>` where the `Any` are `OnceCell<T>` instead of `T`.

This would be flexible, but might be too prescriptive since a user could always store `OnceCell` themselves and just have an extra `.get_or_init(Default::default)` call in the chain.

```rust
static GLOBAL: Global<OnceCell<u32>> = Global::new();

let answer = GLOBAL.get_or_init(Default::default).get_or_init(|| 42);
```